### PR TITLE
KB-H060 No CRLF allowed

### LINKED ISSUE #313: [conan-center] Do not allow CR LF as endline
Related to https://github.com/conan-io/conan-center-index/pull/6055

So far, only ./folly/all/test_package/test_package.cpp uses CRLF and Platform.Interfaces

How to find it?

    cd conan-center-index/recipes && find . -not -type d -exec file "{}" ";" | grep CRLF

### PR DESCRIPTION
**Pull request by uilianries**
closes #313

This hook opens all files in binary more, look for and endline and check which characters are used.

In case CRLF is found, an error is raised, pointing the file path and to replace CRLF by LF only.

### CODE REVIEW

**Review by danimtb — APPROVED** [review]

File: hooks/conan-center.py [hunk 1]
```diff
@@ -56,6 +56,7 @@
              "KB-H055": "SINGLE REQUIRES",
              "KB-H056": "LICENSE PUBLIC DOMAIN",
              "KB-H058": "ILLEGAL CHARACTERS",
+             "KB-H059": "NO CRLF",
```
Comment (uilianries) [inline hooks/conan-center.py:59]:
```suggestion
             "KB-H060": "NO CRLF",
```

File: hooks/conan-center.py [hunk 2]
```diff
@@ -623,6 +624,16 @@ def test(out):
                     out.error("The file '{}' ends with a dot. Please, remove the dot from the end."
                               .format(file, disallowed_chars))
 
+    @run_test("KB-H059", output)
```
Comment (uilianries) [inline hooks/conan-center.py:627]:
```suggestion
    @run_test("KB-H060", output)
```

File: tests/test_hooks/conan-center/test_conan-center.py [hunk 3]
```diff
@@ -1081,3 +1082,15 @@ class AConan(ConanFile):
         output = self.conan(['export', "conanfile.py", 'name/version@user/test'])
         self.assertIn("ERROR: [ILLEGAL CHARACTERS (KB-H058)] The file 'foo.' ends with a dot."
                       " Please, remove the dot from the end.", output)
+
+    def test_no_crlf(self):
+        conanfile = u"from conans import ConanFile\nclass AConan(ConanFile):\n    pass\n"
+
+        tools.save('conanfile.py', content=conanfile)
+        output = self.conan(['export', 'conanfile.py', 'name/version@user/test'])
+        self.assertIn("[NO CRLF (KB-H059)] OK", output)
```
Comment (uilianries) [inline tests/test_hooks/conan-center/test_conan-center.py:1091]:
```suggestion
        self.assertIn("[NO CRLF (KB-H060)] OK", output)
```

File: tests/test_hooks/conan-center/test_conan-center.py [hunk 4]
```diff
@@ -1081,3 +1082,15 @@ class AConan(ConanFile):
...
+
+        with io.open('conanfile.py', 'w', newline='\r\n') as f_handle:
+            f_handle.write(conanfile)
+        output = self.conan(['export', 'conanfile.py', 'name/version@user/test'])
+        self.assertIn("ERROR: [NO CRLF (KB-H059)] The file 'conanfile.py' uses CRLF. Please, replace by LF.", output)
```
Comment (uilianries) [inline tests/test_hooks/conan-center/test_conan-center.py:1096]:
```suggestion
        self.assertIn("ERROR: [NO CRLF (KB-H060)] The file 'conanfile.py' uses CRLF. Please, replace by LF.", output)
```

**Review by uilianries — COMMENTED** [review]
59 is the class name hook

**Review by jgsogo — APPROVED** [review]
Some recipes fail (we don't know if it is because this hook or some other):

```
FAIL: conan export recipes/astc-codec/all astc-codec/cci.20190617@
FAIL: conan export recipes/crunch/all crunch/cci.20190615@
FAIL: conan export recipes/folly/all folly/2019.10.21.00@
FAIL: conan export recipes/folly/all folly/2020.08.10.00@
FAIL: conan export recipes/freetype/all freetype/2.10.4@
FAIL: conan export recipes/freetype/all freetype/2.10.2@
FAIL: conan export recipes/freetype/all freetype/2.10.1@
FAIL: conan export recipes/freetype/all freetype/2.10.0@
FAIL: conan export recipes/guetzli/all guetzli/1.0.1@
FAIL: conan export recipes/harfbuzz/all harfbuzz/2.6.8@
FAIL: conan export recipes/harfbuzz/all harfbuzz/2.7.0@
FAIL: conan export recipes/harfbuzz/all harfbuzz/2.7.1@
FAIL: conan export recipes/harfbuzz/all harfbuzz/2.7.2@
FAIL: conan export recipes/harfbuzz/all harfbuzz/2.7.4@
FAIL: conan export recipes/harfbuzz/all harfbuzz/2.8.0@
FAIL: conan export recipes/harfbuzz/all harfbuzz/2.8.1@
FAIL: conan export recipes/jxrlib/all jxrlib/cci.20170615@
FAIL: conan export recipes/ktx/all ktx/4.0.0@
FAIL: conan export recipes/lodepng/all lodepng/cci.20200615@
FAIL: conan export recipes/meshoptimizer/all meshoptimizer/0.16@
FAIL: conan export recipes/meshoptimizer/all meshoptimizer/0.15@
FAIL: conan export recipes/meshoptimizer/all meshoptimizer/0.14@
FAIL: conan export recipes/msdfgen/all msdfgen/1.9@
FAIL: conan export recipes/opus/all opus/1.3.1@
FAIL: conan export recipes/sdf/all sdf/cci.20141013@
FAIL: conan export recipes/sdl_ttf/all sdl_ttf/2.0.15@
FAIL: conan export recipes/zstd/all zstd/1.4.8@
FAIL: conan export recipes/zstd/all zstd/1.4.9@
FAIL: conan export recipes/zstd/all zstd/1.5.0@
FAIL: conan export recipes/zstd/all zstd/1.3.5@
FAIL: conan export recipes/zstd/all zstd/1.3.8@
FAIL: conan export recipes/zstd/all zstd/1.4.3@
FAIL: conan export recipes/zstd/all zstd/1.4.4@
FAIL: conan export recipes/zstd/all zstd/1.4.5@
FAIL: conan export recipes/zstd/all zstd/1.4.7@
```

Anyway, this is not breaking the world, we can afford it 😄

**Comment by uilianries** [comment]
Last time I checked, only Folly has CRLF. I think those errors are from other hooks

**Comment by uilianries** [comment]
So far, only one file with CRLF

```
$ find . -not -type d -exec file "{}" ";" | grep CRLF
./recipes/folly/all/test_package/test_package.cpp: C source, ASCII text, with CRLF line terminators
```

**Review by danimtb — APPROVED** [review]

**Comment by SpaceIm** [comment]
This hook shouldn't check binary files, this is why it reports errors in several recipes.

**Pull request merged**